src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp

```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
+      raise_or_lower_index(spatial_velocity, spatial_metric);
+
+  const double normal_velocity =
+      get(dot_product(unit_normal, spatial_velocity));
+  const double velocity_squared =
+      get(dot_product(spatial_velocity, spatial_velocity_oneform));
+
+  // Use expression for (hW - 1) with well-behaved Newtonian limit
+  const double h_w_minus_one =
+      get(lorentz_factor) *
+      (get(specific_internal_energy) + get(pressure) / get(rest_mass_density) +
+       get(lorentz_factor) * velocity_squared / (get(lorentz_factor) + 1.0));
+
+  double sound_speed_over_d =
+      sqrt(1.0 - velocity_squared * get(sound_speed_squared) -
+           square(normal_velocity) * (1.0 - get(sound_speed_squared)));
+  sound_speed_over_d *= get(lorentz_factor);
+  sound_speed_over_d = sqrt(get(sound_speed_squared)) / sound_speed_over_d;
+
+  const double specific_enthalpy_times_lorentz_factor =
+      get(specific_enthalpy) * get(lorentz_factor);
+
+  const double h_w_vn_c_over_d = specific_enthalpy_times_lorentz_factor *
+                                 normal_velocity * sound_speed_over_d;
+
+  // set eigenvector with eigenvalue = lapse * Lambda_minus - normal_shift
+  size_t eigenvector_id = 0;
+
+  result(0, eigenvector_id) = 1.0;
+  result(1, eigenvector_id) = h_w_minus_one - h_w_vn_c_over_d;
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = specific_enthalpy_times_lorentz_factor *
+                                    (spatial_velocity_oneform.get(i) -
+                                     sound_speed_over_d * unit_normal.get(i));
+  }
+
+  // set eigenvector with eigenvalue = lapse * Lambda_plus - normal_shift
+  eigenvector_id = Dim + 1;
+
+  result(0, eigenvector_id) = 1.0;
+  result(1, eigenvector_id) = h_w_minus_one + h_w_vn_c_over_d;
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = specific_enthalpy_times_lorentz_factor *
+                                    (spatial_velocity_oneform.get(i) +
+                                     sound_speed_over_d * unit_normal.get(i));
+  }
+
+  // set eigenvector with eigenvalue = lapse * normal_velocity - normal_shift
+  eigenvector_id = Dim;
+
+  result(0, eigenvector_id) = get(kappa_over_density);
+  result(1, eigenvector_id) =
+      get(kappa_over_density) * h_w_minus_one -
+      specific_enthalpy_times_lorentz_factor * get(sound_speed_squared);
+
+  const double prefactor = specific_enthalpy_times_lorentz_factor *
+                           (get(kappa_over_density) - get(sound_speed_squared));
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = prefactor * spatial_velocity_oneform.get(i);
+  }
+
+  // if Dim > 1 make degenerate eigenvectors
+  if (Dim > 1) {
+    const double two_h_w_minus_one =
+        h_w_minus_one + specific_enthalpy_times_lorentz_factor;
+    const auto unit_tangent_oneform =
+        orthonormal_oneform(unit_normal, inv_spatial_metric);
+    const auto degenerate_eigenvector =
```

Comment:
[nitpick] I suggest a more "active" name to clarify what this is doing... perhaps "set_degenerate_eigenvector"? ... up to you on this one.